### PR TITLE
please - a convenience layer around the nix tooling

### PR DESCRIPTION
**Pull request by gilligan**
```
Usage: please <command> [args]

 build [artifact]    -- build an artifact
 completions         -- output completion script
 doctor              -- perform sanity checks
 init                -- configure initial setup
 install [artifact]  -- install an artifact
 list                -- list artifacts and tests
 run-test [test]     -- run a test
 run-vm [test]       -- run an interactive vm of a test
 shell [artifact]    -- enter a dev shell for an artifact

In order to enable context-sensitive completions (bash only!) run:

  $ source <(please completions)

You should add this to your init scripts.
```

This PR adds a command *please* which acts as a convenience layer for using what this repo provides. It's basically a very thin layer that just invokes various nix commands but enables users to get going without any prior knowledge of Nix. 

`please doctor` and `please init` can be used to get all configuration done and replaces the manual steps outiined in the README.md - *caveat* if there already is a `.config/nix/nix.conf` file, the file will not be changed as this is too error-prone.

bash completions are provided for all commands. They rely on the sanitized output of `jobset.nix` and should make it much easier to find/build/install what users might be looking for.

**Note**
This PR also extracts the previously unexposed versions array to `./contrail-versions.nix` in order to make it easy to retrieve the currently supported contrail versions.

### TODOs

- [x] `./please init` should NOT bail out when `nix` isn't already installed
- [x] `/.please init` should INSTALL nix
- [x] Make sure it is usable even when unicode characters cannot be displayed
- [x] Add some tests
  - [x] test `init` and `doctor`
  - [x] make sure the expressions in `get-attr.nix` actually work / are not out of sync
- [x] Update documentation (README.md)
Regarding the order, I think please shell and please run-test should be moved to the end of the please section.
- [X] I think the table of content is too big (we don't need to detail each please section)
- [X] In the introdution, we could say we provide a thin layer on top of Nix and Nix user can find the nix command executed in the output of each please section.
- [x] What about propagating any additional please argument to the executed nix command?
- [x] You can replace nixpkgs-contrail by nixpkgs-tungsten everywhere :)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,141 +33,203 @@ The only prerequisite and hard dependency of **nixpkgs-contrail** is
...
+ init                -- configure initial setup
+ install [artifact]  -- install an artifact
+ list                -- list artifacts and tests
+ run-test [test]     -- run a test
+ run-vm [test]       -- run an interactive vm of a test
+ shell [artifact]    -- enter a dev shell for an artifact
+
+In order to enable context-sensitive completions (bash only!) run:
+
+  $ source <(please completions)
```
Comment (nlewo) [inline README.md]:
```suggestion
  $ source <(./please completions)
```

File: please [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+_sourceNix() {
+    NIX_SH="$HOME/.nix-profile/etc/profile.d/nix.sh"
+    test -f "$NIX_SH" && source "$NIX_SH"
+}
+
+_addCacheConfig() {
+    if test -f ~/.config/nix/nix.conf
+    then
+        log "$HOME/.config/nix/nix.conf exists. Please follow the instructions from the README"
```
Comment (nlewo) [inline please:67]: Wouyld be nice to provide the README section

**Review by nlewo — COMMENTED** [review]
Maybe it would be nice to add an `uninstall` command.

**Comment by nlewo** [comment]
Note I still have to review the README.

File: README.md [hunk 3]
```diff
@@ -31,141 +33,203 @@ The only prerequisite and hard dependency of **nixpkgs-contrail** is
...
+
+You should add this to your init scripts.
+
 ```
 
-For more detailed information about Nix please refer to [https://nixos.org/nix](https://nixos.org/nix/).
+The `./please` script provides a convenience layer for performing the most typical actions such as building or installing packages provided through **nixpkgs-contrail**.
+Experienced Nix users are more likely to use nix tooling directly, for everyone
```
Comment (nlewo) [inline README.md]:
```suggestion
Experienced Nix users are more likely to use Nix tooling directly, for everyone 
```

File: README.md [hunk 4]
```diff
@@ -31,141 +33,203 @@ The only prerequisite and hard dependency of **nixpkgs-contrail** is
...
+
+In order to install any **nixpkgs-contrail** package you can use `./please install`:
 
 ```
-$ nix-build -A contrail32.test.allInOne.driver
-$ QEMU_NET_OPTS="hostfwd=tcp::2222-:22" result/bin/nixos-run-vms
-$ ssh -p 2222 root@localhost
-Password: <ENTER>
+$ ./please install contrail50.apiServer 
+[please]: Running "nix-env -f default.nix -iA contrail50.apiServer"
```
Comment (nlewo) [inline README.md]: It would be better to install a tool instead of a Contrail component.

File: README.md [hunk 5]
```diff
@@ -31,141 +33,203 @@ The only prerequisite and hard dependency of **nixpkgs-contrail** is
...
+These tests are executed in virtual machines running in QEMU and can be executed
+using the `run-test` command.
 
 ```
-$ nix-shell -A contrail32.control
+$ ./please run-test contrail50.test.tcpFlows
+[please]: Running "nix-build default.nix -A contrail50.test.tcpFlows"
 ```
+Apart from generating a lot of output on the terminal, each test execution will also
+ceate a `result` output link containing a `log.html` file which contains a pretty-printed
```
Comment (nlewo) [inline README.md]:
```suggestion
create a `result` output link containing a `log.html` file which contains a pretty-printed 
```

**Review by nlewo — COMMENTED** [review]
- Regarding the order, I think `please shell` and `please run-test ` should be moved to the end of the please section.
- I think the table of content is too big (we don't need to detail each please section)
- In the introdution, we could say we provide a thin layer on top of Nix and Nix user can find the nix command executed in the output of each please section.
- What about propagating any additional please argument to the executed nix command?
- You can replace nixpkgs-contrail by nixpkgs-tungsten everywhere :)

File: please (hunk 2 above)
Comment (gilligan) [reply please:67]: Oh, Yes. Especially now that I realized that I have actually removed that section from the README so I have to fix this. Thanks for noticing.

**Comment by gilligan** [comment]
> What about propagating any additional please argument to the executed nix command?

@nlewo At first I thought this would be a good idea but actually I think prefer to not implement this. At least not right now. I actually limited the `please` commands so that they only work with exactly one argument each. In that sense it maybe isn't an "abstraction" but an encapsulation but my focus was on making sure that people can use this to get things done and without doing any mistakes.

Long story short - for *this* PR I would prefer to not add this but i'm open to getting back to this and adding it at a later point. Is that OK with you?

File: README.md (hunk 4 above)
Comment (gilligan) [reply README.md]: Yup - so i went from `gremlinConsole` instead

**Comment by nlewo** [comment]
Thx!
